Add support for empty topic names.

prakhar1144 opened the pull request:
[CZO - Design Discussion](https://chat.zulip.org/#narrow/channel/101-design/topic/general.20chat.20.2F.20default.20topic)

[CZO - API Design Discussion](https://chat.zulip.org/#narrow/channel/378-api-design/topic/general.20chat)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
It'd be great to receive an initial review on this. 

* We'll stash all the commits in one, kept it separate for the ease of review and re-iterating on different chunks.

* Currently, I have ignored tests and documentation. -- that'll be quick to do if implementation looks fine.

---

The only left part is related to narrows:
* /messages/matches_narrow
* `match_subject` field in `GET /messages` response.

timabbott commented on zerver/lib/topic.py:
```diff
@@ -321,3 +321,9 @@ def participants_for_topic(realm_id: int, recipient_id: int, topic_name: str) ->
+def maybe_rename_general_chat_to_empty_topic(topic_name: str) -> str:
+    if topic_name == "general chat":
+        topic_name = ""
+    return topic_name
```
I'm not sure we want this behavior here; it should be in the `do_send_message` code path only, only for clients that don't support the `""` topic. We don't want to impact clients that do understand the modern general chat format and want to send to "general chat" as a topic, if we can avoid it.

timabbott replied:
Hmm, I guess what's awkward is that the "send message" API request code path does not at present accept an `empty_topic_name` client capability parameter. But probably we could make it do that?

timabbott replied:
Having now read the branch more fully, I think you've taken the right approach; too many other code paths need to deal with this for us to avoid sacrificing "general chat" as a valid topic during the transitional period. 

We will probably want to move relatively quickly on this transition, though, in terms of adding explicit mobile app support for the client capability, though, since this does feel like a messy wart to live with.

timabbott commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -61,6 +61,8 @@ def make_request(self, base_url: str, event: dict[str, Any], realm: Realm) -> Re
             apply_markdown=False,
             client_gravatar=False,
+            # TODO: Discuss in CZO.
+            allow_empty_topic_name=False,
```
I think we probably want `True` here -- we have no reason to believe outgoing webhooks would fail to process `""` as a topic, and it seems better to keep things simple for them. We do need to be careful to make sure `""` round-trips properly without getting converted to "general chat" for outgoing webhook bots that want to trigger a reply, but that should be OK.

prakhar1144 replied:
Used `True`

timabbott commented on zerver/lib/message_cache.py:
```diff
@@ -219,6 +222,16 @@ def finalize_payload(
+            and obj["subject"] == ""
+            and not allow_empty_topic_name
+        ):
+            obj["subject"] = "general chat"
```
So I had a thought on this -- we could perhaps have `topic_name` be available and have a value of `""`, and a legacy/deprecated `"subject"` field present whose value is `"general chat"` here. 

The idea being that a modern client would ignore `subject` and just parse the `topic_name` field, and thus we'd be able to remove the legacy `subject` name along with this transition. And the `allow_empty_topic_name` client capability could then just have a meaning of "skip sending the legacy `subject` field". 

Worth discussing in the API design channel, but this might be a pretty compelling way to move forward.

timabbott replied:
Having now read the rest of the branch, I think the essential idea of that proposal is a good idea, but probably something we can do in a follow-up. Basically, we'd introduce a `topic_name` field on `Message` that equals `subject`, except that it doesn't support `!allow_empty_topic_name`, under the assumptions that clients that migrate to reading it should post-date that API change anyway. And in doing so, we'd at least set ourselves up for the final end of the original "subject" name for topics (we renamed it in the UI in like 2013).

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -1286,19 +1287,21 @@ def process_deletion_event(event: Mapping[str, Any], users: Iterable[int]) -> No
             if not client.accepts_event(event):
                 continue
 
+            deletion_event = dict(event)
```
Why do we need this copy in addition to the `.copy` steps below?

prakhar1144 replied:
We have a `continue` statement case before that `.copy`. -- `Mapping` is read only type, we can't alter `event` directly.

prakhar1144 replied:
One improvement we can make in a separate small PR for:
```py
            for message_id in deletion_event["message_ids"]:
                compatibility_event = deletion_event.copy()
                compatibility_event["message_id"] = message_id
                del compatibility_event["message_ids"]
                client.add_event(compatibility_event)
```
to 
```py
            message_ids = deletion_event["message_ids"]
            del deletion_event["message_ids"]
            for message_id in message_ids:
                deletion_event["message_id"] = message_id
                client.add_event(deletion_event)
```

timabbott replied:
I'm pretty sure that change would be a serious bug. If one just edits the same object, inserting it into different queues, then one will end up with many pointers to the same object (all with the last message ID at the end of the loop).  

I'm sure some refactor works here, but the one you propose seems dangerous.

prakhar1144 replied:
Made a small change to copy only if `topic == "" and not client.empty_topic_name`

```py
def process_deletion_event(event: Mapping[str, Any], users: Iterable[int]) -> None:
    for user_profile_id in users:
        for client in get_client_descriptors_for_user(user_profile_id):
            if not client.accepts_event(event):
                continue

            deletion_event = event
            if deletion_event.get("topic") == "" and not client.empty_topic_name:
                deletion_event = dict(event)
                deletion_event["topic"] = Message.EMPTY_TOPIC_FALLBACK_NAME

            # For clients which support message deletion in bulk, we
            # send a list of msgs_ids together, otherwise we send a
            # delete event for each message.  All clients will be
            # required to support bulk_message_deletion in the future;
            # this logic is intended for backwards-compatibility only.
            if client.bulk_message_deletion:
                client.add_event(deletion_event)
                continue

            for message_id in deletion_event["message_ids"]:
                compatibility_event = dict(deletion_event)
                compatibility_event["message_id"] = message_id
                del compatibility_event["message_ids"]
                client.add_event(compatibility_event)
```

timabbott replied:
OK, this works.

timabbott commented on zerver/views/message_edit.py:
```diff
@@ -99,6 +107,7 @@ def get_message_edit_history(
     user_profile: UserProfile,
     *,
     message_id: PathOnly[NonNegativeInt],
+    allow_empty_topic_name: Json[bool] = False,
```
We might be able to skip adding this parameter. I suspect that no client app uses this API endpoint -- the only thing they use is the edit history data that comes in `GET /messages` responses and events.

I guess the benefit to doing so is that that we'd save ourselves a future transition of changing the default from False => True here.

prakhar1144 replied:
>  I suspect that no client app uses this API endpoint

Terminal uses this endpoint.

See: https://chat.zulip.org/#narrow/channel/378-api-design/topic/edit.20history/near/1933192

timabbott replied:
OK, thanks for checking, nothing can be readily improved here, then.

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -1637,6 +1637,24 @@ def process_stream_typing_notification_event(event: Mapping[str, Any], users: It
+                ):
+                    event_copy["message_details"][message_id]["topic"] = "general chat"
+
+            client.add_event(event_copy)
```
This can probably be optimized to skip the copying step in the (default for the web app) case where `client.empty_topic_name`. I think generally we do want to write this event_queue.py code that way, since it is performa-sensitive code, though this particular event is rare enough that it likely doesn't matter.

Also, rather than hardcoding "general chat" here, we probably want to be using a global, like `Message.EMPTY_TOPIC_FALLBACK_NAME`, to avoid having lots of hardcoded references to that topic name..

prakhar1144 replied:
> Also, rather than hardcoding "general chat" here, we probably want to be using a global, like Message.EMPTY_TOPIC_FALLBACK_NAME

Done

> This can probably be optimized to skip the copying step in the case where client.empty_topic_name

Done

prakhar1144 replied:
> This can probably be optimized to skip the copying step in the case where client.empty_topic_name

Based on this feedback, did some optimization in `process_user_topic_event` and `process_stream_typing_notification_event`.

timabbott replied:
I posted some ideas for how to do that more nicely in another comment; this code can probably do the same approach of doing the `deepcopy` once up front.

prakhar1144 replied:
Made the changes to perform deepcopy once.

prakhar1144 commented:
API Docs part is done. Working on tests.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -1747,7 +1747,7 @@ def check_message(
             assert sender.bot_type == sender.OUTGOING_WEBHOOK_BOT
 
-        if realm.mandatory_topics and topic_name == "(no topic)":
+        if realm.mandatory_topics and topic_name == "":
```
I think we might need to forbid both `""` and `"(no topic)"` during a transitional period. Ideally, we'd be able to merge the whole server part of this PR without any user-facing impact (by just not enabling the client capability), and we don't want to allow the "(no topic)" topic until we're ready to cut over to the feature being fully supported.

prakhar1144 replied:
> I think we might need to forbid both "" and "(no topic)"

Yeah, makes sense. Made the changes.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -14116,6 +14116,13 @@ paths:
+                    - `empty_topic_name`: Boolean for whether the client supports empty string as
+                      a topic name in `register` response or events involving topic names.
+                      Clients that don't support this client capability receive `"general chat"`
+                      as the topic name replacing `""`.
```
Proposed edit:

> Boolean for whether the client supports processing the empty string as a topic name. Clients not declaring this capability will be sent `"general chat"` instead of the empty string wherever topic names appear in the `register` response or events involving topic names.

> **Changes"": New in .... Previously, the empty string was not a valid topic.

(Small note: When we add an API feature for what to call general chat, we will want to update this language to reference that).

prakhar1144 replied:
Done. 

Replaced `"general chat"` with _**"the value of `realm_empty_topic_display_name` field in the [`POST /register`](/api/register-queue) response"**_

timabbott commented on zerver/tests/test_message_move_topic.py:
```diff
@@ -93,19 +93,6 @@ def test_propagate_invalid(self) -> None:
-                "topic": " ",
-            },
-        )
-        self.assert_json_error(result, "Topic can't be empty!")
```
We might want to preserve this test scenario (renaming the test), and just check it does something different, naming that it'll have moved the message to the `""` topic, despite the space in what the client snet.

prakhar1144 replied:
Renamed it to `test_edit_message_empty_topic_with_extra_space` and did the necessary tweaks.

timabbott commented on zerver/tests/test_message_send.py:
```diff
@@ -3358,7 +3342,7 @@ def test_no_topic_message(self) -> None:
         client = make_client(name="test suite")
         stream = get_stream("Denmark", realm)
-        topic_name = "(no topic)"
+        topic_name = ""
```
As noted above, we probably want to split this into two tests, one for "(no topic)" legacy behavior, and one for the new `""`.

prakhar1144 replied:
Done

timabbott commented on zerver/tests/test_message_topics.py:
```diff
@@ -333,3 +335,55 @@ def test_topic_delete_timeout(self) -> None:
             self.assertFalse(result_dict["complete"])
+
+
+class EmptyTopicNameTest(ZulipTestCase):
```
This test is great.

timabbott commented on zerver/tests/test_message_topics.py:
```diff
@@ -333,3 +335,55 @@ def test_topic_delete_timeout(self) -> None:
+
+        self.send_stream_message(iago, "Denmark", topic_name="general chat")
+        events = client.event_queue.contents()
+        self.assertEqual(events[1]["message"]["subject"], "general chat")
```
This test code could probably use `EMPTY_TOPIC_FALLBACK_NAME`

prakhar1144 replied:
Yeah, replaced them with `Message.EMPTY_TOPIC_FALLBACK_NAME`

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -6793,6 +6793,19 @@ paths:
+            Whether the topic names (`subject` field, and `prev_topic`, `topic` in `edit_history` objects)
+            in the messages returned can be empty string.
+
+            If `false`, `"general chat"` is returned replacing `""` as the topic name.
```
> Whether the client supports processing the empty string as a topic in the topic name fields in the returned data, including in returned `edit_history` data. 

> If `false`, the server will use `"general chat"` instead of `""` to represent the empty string topic in its response.

And add to the changes entry:

> Previously, the empty string was not a valid topic.

prakhar1144 replied:
Done

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -1222,20 +1224,11 @@ def get_client_payload(
-            and wide_dict["subject"] == ""
-        ):
-            message_dict = message_dict.copy()
-            message_dict["subject"] = Message.EMPTY_TOPIC_FALLBACK_NAME
```
I'm a little confused by why the second commit removes this.

prakhar1144 replied:
`get_client_payload` makes a call to `MessageDict.finalize_payload` which performs this logic. We can stash the two commits -- kept it separate as the changes will be a lot to review in one commit.

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -1622,6 +1622,21 @@ def process_user_group_name_update_event(event: Mapping[str, Any], users: Iterab
+
+            event_copy = dict(event)
+            event_copy["topic_name"] = Message.EMPTY_TOPIC_FALLBACK_NAME
+            client.add_event(event_copy)
```
The optimization that would work here is something like:
```
if event.get("topic_name") == "":
    empty_topic_name_fallback_event = dict(event)
    empty_topic_name_fallback_event["topic_name"] = Message.EMPTY_TOPIC_FALLBACK_NAME
else: 
    empty_topic_name_fallback_event = event
```

and then inside the loop, just check `client.empty_topic_name` and do `add_event` on either `event` or `empty_topic_name_fallback_event`.

Then we only construct the fallback event once per event, not once per client without the client capability (which will be all of them).

prakhar1144 replied:
Made the changes

timabbott commented on api_docs/changelog.md:
```diff
@@ -34,7 +34,8 @@ format used by the Zulip server that they are interacting with.
   in the `update_message` event type, `topic_name` field in the `user_topic`
-  event type will have the value `"general chat"` replacing `""` for channel messages.
+  event type, `topic` field in the `typing` event type will have the value
+  `"general chat"` replacing `""` for channel messages.
```
I think this might have been better written with a 2-layer bulleted list. First, an outer one giving the context on the change, and then the inner one just listing the fields.

prakhar1144 replied:
<img width="686" alt="Screenshot 2024-12-19 at 10 24 03 PM" src="https://github.com/user-attachments/assets/560e0943-bc8f-474b-b1f2-2e69da7e3a05" />

timabbott replied:
Yeah, this is great.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -7308,13 +7321,19 @@ paths:
                                 prior to this edit event.
+
+                                If `allow_empty_topic_name` parameter is `false`,
+                                the `""` value is replaced with `"general chat"`.
```
I don't think we need this repetition here; the behavior is explained clearly in the description of the parameter.

prakhar1144 replied:
Removed

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -15055,6 +15055,15 @@ paths:
+
+                                    **Changes**: Before 10.0 (feature level 331), `empty_topic_name`
+                                    client capability didn't exist and empty string as the topic name for
+                                    channel messages wasn't allowed.
```
For the last two paragraphs, maybe something like this:

> Note that the `""` topic may have been rewritten by the server to "general chat" depending on the value of the `"empty_topic_name"` client capability.

> **Changes**: The `empty_topic_name` client capability is new in Zulip 10.0 (feature level 331).

prakhar1144 replied:
Done

timabbott commented on zerver/lib/events.py:
```diff
@@ -1972,3 +1974,8 @@ def post_process_state(
+    if not allow_empty_topic_name and "user_topics" in ret:
+        for user_topic in ret["user_topics"]:
+            if user_topic["topic_name"] == "":
+                user_topic["topic_name"] = Message.EMPTY_TOPIC_FALLBACK_NAME
```
I don't understand this block -- can you explain why we need a block here? Do the tests fail if this code is deleted?

I'd have expected that this would be handled just via the earlier code; this post-processing function is meant to be only for things like translating a `dict` to ` list` for the API... and I'd have thought the reformatting would be done earlier.

Since we know the event queue will have the same client capability value as the `fetch_initial_state_data` caller, I think just changing the initial population code should be sufficient, and `apply_events` will do the right thing via only processing topics in the same namespace,

prakhar1144 replied:
That'll also work but it'll lead to more compatibility code additions within other lib functions.

For example, if we add this logic in `get_user_topics` (the initial population code) then it'll affect `get_topic_mutes` which is used in `bulk_do_set_user_topic_visibility_policy` -- so, we'll have to add some compatibility code in `bulk_do_set_user_topic_visibility_policy` which is used at a lot of places and will lead to a lot of changes.

So, I think just adding it in the last stage of the flow is the right approach ?

timabbott replied:
OK, that's probably fine.

timabbott commented:
I prefer this split out as written, so don't squash the commits.

prakhar1144 commented on zerver/models/messages.py:
```diff
@@ -157,6 +157,8 @@ class Message(AbstractMessage):
 
     DEFAULT_SELECT_RELATED = ["sender", "realm", "recipient", "sending_client"]
 
+    EMPTY_TOPIC_FALLBACK_NAME = "general chat"
```
Comment added.

prakhar1144 commented:
Updates in the recent push:
* Addressed review comments
* Added first commit as per yesterday's [discussion in CZO.](https://chat.zulip.org/#narrow/channel/378-api-design/topic/general.20chat/near/2009033)
* Wrote commit messages.

prakhar1144 commented:
[Last Update](https://github.com/zulip/zulip/pull/32330#issuecomment-2555113031)

* In the recent push, added code in the 3rd commit so that narrowing to "general chat" works properly.

  <details><summary>git diff</summary>
  <p>
  
  ```diff
  diff --git a/zerver/lib/narrow.py b/zerver/lib/narrow.py
  index 437914cfce..a7698cc967 100644
  --- a/zerver/lib/narrow.py
  +++ b/zerver/lib/narrow.py
  @@ -49,6 +49,7 @@ from zerver.lib.streams import (
       get_stream_by_narrow_operand_access_unchecked,
       get_web_public_streams_queryset,
   )
  +from zerver.lib.topic import maybe_rename_general_chat_to_empty_topic
   from zerver.lib.topic_sqlalchemy import (
       get_followed_topic_condition_sa,
       get_resolved_topic_condition_sa,
  @@ -900,6 +901,20 @@ def can_narrow_define_conversation(narrow: list[NarrowParameter]) -> bool:
       return False
  
  
  +def update_narrow_terms_containing_empty_topic_fallback_name(
  +    narrow: list[NarrowParameter] | None,
  +) -> list[NarrowParameter] | None:
  +    if narrow is None:
  +        return narrow
  +
  +    for term in narrow:
  +        if term.operator == "topic":
  +            term.operand = maybe_rename_general_chat_to_empty_topic(term.operand)
  +            break
  +
  +    return narrow
  +
  +
   # This function implements the core logic of the `with` operator,
   # which is designed to support permanent links to a topic that
   # robustly function if the topic is moved.
  diff --git a/zerver/tests/test_message_topics.py b/zerver/tests/test_message_topics.py
  index 4970102a15..cc0ba8d9f7 100644
  --- a/zerver/tests/test_message_topics.py
  +++ b/zerver/tests/test_message_topics.py
  @@ -405,6 +405,7 @@ class EmptyTopicNameTest(ZulipTestCase):
           }
           result = self.client_get("/json/messages", params)
           data = self.assert_json_success(result)
  +        self.assert_length(data["messages"], 2)
           for message in data["messages"]:
               self.assertEqual(message["subject"], Message.EMPTY_TOPIC_FALLBACK_NAME)
  
  @@ -414,9 +415,28 @@ class EmptyTopicNameTest(ZulipTestCase):
           }
           result = self.client_get("/json/messages", params)
           data = self.assert_json_success(result)
  +        self.assert_length(data["messages"], 2)
           for message in data["messages"]:
               self.assertEqual(message["subject"], "")
  
  +        get_params = {
  +            "allow_empty_topic_name": "false",
  +            "anchor": "newest",
  +            "num_before": 2,
  +            "num_after": 0,
  +            "narrow": orjson.dumps(
  +                [
  +                    {"operator": "channel", "operand": "Denmark"},
  +                    {"operator": "topic", "operand": Message.EMPTY_TOPIC_FALLBACK_NAME},
  +                ]
  +            ).decode(),
  +        }
  +        result = self.client_get("/json/messages", get_params)
  +        data = self.assert_json_success(result)
  +        self.assert_length(data["messages"], 2)
  +        for message in data["messages"]:
  +            self.assertEqual(message["subject"], Message.EMPTY_TOPIC_FALLBACK_NAME)
  +
           # Fetch using `/messages/{message_id}` endpoint.
           result = self.client_get(
               f"/json/messages/{first_message_id}", {"allow_empty_topic_name": "false"}
  diff --git a/zerver/views/message_fetch.py b/zerver/views/message_fetch.py
  index e9f91bfbb2..c1654c5c05 100644
  --- a/zerver/views/message_fetch.py
  +++ b/zerver/views/message_fetch.py
  @@ -25,6 +25,7 @@ from zerver.lib.narrow import (
       is_spectator_compatible,
       is_web_public_narrow,
       parse_anchor_value,
  +    update_narrow_terms_containing_empty_topic_fallback_name,
       update_narrow_terms_containing_with_operator,
   )
   from zerver.lib.request import RequestNotes
  @@ -145,6 +146,7 @@ def get_messages_backend(
           anchor = parse_anchor_value(anchor_val, use_first_unread_anchor_val)
  
       realm = get_valid_realm_from_request(request)
  +    narrow = update_narrow_terms_containing_empty_topic_fallback_name(narrow)
       narrow = update_narrow_terms_containing_with_operator(realm, maybe_user_profile, narrow)
  
       num_of_messages_requested = num_before + num_after
  
  ```
  
  </p>
  </details>

timabbott commented on zerver/tests/test_message_topics.py:
```diff
@@ -756,3 +757,39 @@ def test_get_message_edit_history(self) -> None:
+                get_client("website"),
+                client_capabilities=ClientCapabilities(
+                    empty_topic_name=True, notification_settings_null=False
+                ),
```
This sort of test should probably pass a short `fetch_event_types` to make the test do ~20x less work; too many `do_events_register` calls without that slows down the test suite.

prakhar1144 replied:
PR #34201

timabbott commented:
This looks great, rebased to resolve merge conflicts and marking to merge once CI passes. See a small improvement to be made to the tests noted above. I'll post in "api design" about next steps.
